### PR TITLE
[SGD] v2 Backend specific checkpoint conversion

### PR DESCRIPTION
**Pull request by amogkam**
Adds functionality for backend specific checkpoints converting and unconverting. When checkpoints are collected on the workers, they are first go through the convert function before being sent to the driver. On the driver side, they are unconverted and then stored as checkpoints.

This is useful for Torch where we would run into deserialization errors if we try to checkpoint a GPU model on the driver that only has CPUs. Instead we want to convert the checkpoint to bytes on the worker side, and then convert back to dict on the driver side. 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/util/sgd/v2/backends/backend.py [hunk 1]
```diff
@@ -88,6 +88,7 @@ def initialize_session(world_rank, train_func, checkpoint):
                     training_func=train_func,
                     world_rank=world_rank,
                     checkpoint=checkpoint,
+                    process_checkpoint_func=self._backend.convert_checkpoint,
```
Comment (amogkam) [inline python/ray/util/sgd/v2/backends/backend.py:91]:
```suggestion
                    convert_checkpoint_func=self._backend.convert_checkpoint,
```

File: python/ray/util/sgd/v2/backends/torch.py [hunk 2]
```diff
@@ -140,3 +141,26 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+        """Converts stream of bytes back to a checkpoint Dict.
+
+        Converts to GPU if ``to_gpu`` is True and CUDA is available.
+        """
+        _buffer = io.BytesIO(data)
+        to_gpu = to_gpu and torch.cuda.is_available()
+        state_dict = torch.load(
+            _buffer,
+            map_location=("cpu" if not to_gpu else
+                          lambda storage, loc: storage.cuda()))
```
Comment (matthewdeng) [inline python/ray/util/sgd/v2/backends/torch.py:165]: What's the benefit of loading this in GPU memory on the driver?

File: python/ray/util/sgd/v2/backends/backend.py [hunk 3]
```diff
@@ -88,6 +88,7 @@ def initialize_session(world_rank, train_func, checkpoint):
                     training_func=train_func,
                     world_rank=world_rank,
                     checkpoint=checkpoint,
+                    convert_checkpoint_func=self._backend.convert_checkpoint,
                     detailed_autofilled_metrics=use_detailed_autofilled_metrics
                 )
```
Comment (matthewdeng) [inline python/ray/util/sgd/v2/backends/backend.py:93]: Do we need to `convert` the loaded checkpoint on the driver and `unconvert` on the worker session?

File: python/ray/util/sgd/v2/backends/torch.py [hunk 4]
```diff
@@ -140,3 +141,26 @@ def on_shutdown(self, worker_group: WorkerGroup,
...
+
+    def convert_checkpoint(self, checkpoint: Dict) -> Any:
+        """Convert checkpoint to stream of bytes.
+
+        This prevents deserialization errors if driver is CPU only but
+        workers are on GPU.
+        """
+        _buffer = io.BytesIO()
+        torch.save(checkpoint, _buffer)
+        return _buffer.getvalue()
```
Comment (matthewdeng) [inline python/ray/util/sgd/v2/backends/torch.py:153]: Is there a more specific type hint that can be used for this return type?

File: python/ray/util/sgd/v2/backends/backend.py [hunk 5]
```diff
@@ -329,6 +333,12 @@ def on_shutdown(self, worker_group: WorkerGroup,
                     backend_config: BackendConfig):
         raise NotImplementedError
 
+    def convert_checkpoint(self, checkpoint: Dict) -> Any:
+        pass
+
+    def unconvert_checkpoint(self, data: Any, to_gpu: bool = False) -> Dict:
+        pass
+
```
Comment (matthewdeng) [inline python/ray/util/sgd/v2/backends/backend.py:341]: Nit: I think the naming for these methods is quite confusing (maybe "convert" is a bit generic and does not immediately convey that this is for network transfer). What do you think about `serialize`/`deserialize` or `encode`/`decode`?

File: python/ray/util/sgd/v2/backends/torch.py (hunk 2 above)
Comment (amogkam) [reply python/ray/util/sgd/v2/backends/torch.py:165]: The idea here was that if the user checkpoints a GPU model, we should still keep it as GPU unless no GPUs are available on the driver machine.

Comment (amogkam) [reply python/ray/util/sgd/v2/backends/torch.py:165]: This brings up a good point though which I didn't consider- what if the checkpointed model on the worker side is on CPU? Then we should still keep it as CPU on the driver side even if there's GPU available. Currently though there's no easy way to identify this from the bytes.

File: python/ray/util/sgd/v2/backends/backend.py (hunk 5 above)
Comment (richardliaw) [reply python/ray/util/sgd/v2/backends/backend.py:341]: +1

File: python/ray/util/sgd/v2/backends/torch.py (hunk 4 above)
Comment (richardliaw) [reply python/ray/util/sgd/v2/backends/torch.py:153]: Should these be class methods? they don't rely on self

**Comment by matthewdeng** [comment]
Fixed in https://github.com/ray-project/ray/pull/20333.

**Pull request closed without merging**